### PR TITLE
Add Type Convert

### PR DESCRIPTION
**Pull request by lightszero**
fixed https://github.com/neo-project/neo-devpack-dotnet/issues/195

### CODE REVIEW

File: tests/Neo.Compiler.MSIL.UnitTests/TestClasses/Contract_TypeConvert.cs [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+    class Contract_TypeConvert : SmartContract.Framework.SmartContract
+    {
+        //Integer = 0x21,
+        //ByteArray = 0x28,
+        [OpCode(SmartContract.Framework.OpCode.CONVERT, "0x28")]
+        public extern static byte[] ConvertToByteArray(object from);
+
+
+        [OpCode(SmartContract.Framework.OpCode.CONVERT, "0x21")]
+        public extern static BigInteger ConvertToInteger(byte[] from);
```
Comment (Tommo-L) [inline tests/Neo.Compiler.MSIL.UnitTests/TestClasses/Contract_TypeConvert.cs]: Could we still use the `AsByteArray` as the function name?

Comment (shargon) [reply tests/Neo.Compiler.MSIL.UnitTests/TestClasses/Contract_TypeConvert.cs]: Maybe we should move these to `SmartContract.Framework`

Comment (lightszero) [reply tests/Neo.Compiler.MSIL.UnitTests/TestClasses/Contract_TypeConvert.cs]: if we do not  care about old contract code,we can direct change AsByteArray.

Comment (Tommo-L) [reply tests/Neo.Compiler.MSIL.UnitTests/TestClasses/Contract_TypeConvert.cs]: Just ignore the old contract, as the previous version could not invoke contract successful.

Comment (lightszero) [reply tests/Neo.Compiler.MSIL.UnitTests/TestClasses/Contract_TypeConvert.cs]: ok，let us add more convert code in framework and compiler

**Comment by lightszero** [comment]
I had do some fix

now 
new byte[0]  =>ByteArray
new byte[3]{1,2,3}=>ByteArray
true        =>Boolean
1=>Integer
2=>Integer
17          =>Integer

const value will got right type without convert.

**Comment by lightszero** [comment]
Helper.AsByteArray() fixed.

File: tests/Neo.Compiler.MSIL.UnitTests/UnitTest_TypeConvert.cs [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+        [TestMethod]
+        public void UnitTest_TestTypeConvert()
+        {
+            var testengine = new TestEngine();
+            testengine.AddEntryScript("./TestClasses/Contract_TypeConvert.cs");
+            var result = testengine.ExecuteTestCaseStandard("testType");
+
+            //test 0,1,2
+            Assert.IsTrue(result.TryPop(out Array arr));
+            Assert.IsTrue(arr[0].Type == StackItemType.Integer);
```
Comment (shargon) [inline tests/Neo.Compiler.MSIL.UnitTests/UnitTest_TypeConvert.cs]: Also we should check the values

Comment (lightszero) [reply tests/Neo.Compiler.MSIL.UnitTests/UnitTest_TypeConvert.cs]: sure

**Comment by lightszero** [comment]
@shargon has a format problem,fix it.
@Tommo-L what do you think?should we merge this?

**Comment by Tommo-L** [comment]
I'll review it today.

File: src/Neo.Compiler.MSIL/MSIL/Conv_Common.cs [hunk 3]
```diff
@@ -93,9 +93,29 @@ private NeoCode _Convert1by1(VM.OpCode code, OpCode src, NeoMethod to, byte[] da
...
+            }
+        }
+
+        private void _ConvertPushBoolean(bool b, OpCode src, NeoMethod to)
+        {
+            if (!b)
+                _Convert1by1(VM.OpCode.PUSH0, src, to);
+            else
+                _Convert1by1(VM.OpCode.PUSH1, src, to);
+            _Insert1(VM.OpCode.CONVERT, "", to, new byte[1] { (byte)VM.Types.StackItemType.Boolean });
```
Comment (Tommo-L) [inline src/Neo.Compiler.MSIL/MSIL/Conv_Common.cs]: Repeated?

File: src/Neo.SmartContract.Framework/Helper.cs [hunk 4]
```diff
@@ -5,59 +5,62 @@ namespace Neo.SmartContract.Framework
...
-        //}
+
 
         /// <summary>
         /// Converts BigInteger to byte[]. No guarantees are assumed regarding BigInteger working range.
         /// Examples: 10 -> [0x0a]; 10 -> [0x0a00]; -128 -> [0x80]; -128 -> [0x80ff]; 0 -> []; 0 -> [0x00]; 255 -> [0xff00]
         /// </summary>
-        [Script]
+        [OpCode(OpCode.CONVERT, StackItemType_ByteArray)]
         public extern static byte[] AsByteArray(this BigInteger source);
```
Comment (Tommo-L) [inline src/Neo.SmartContract.Framework/Helper.cs]: Maybe we need to update the comments later.

File: src/Neo.Compiler.MSIL/MSIL/Conv_Common.cs (hunk 3 above)
Comment (Tommo-L) [reply src/Neo.Compiler.MSIL/MSIL/Conv_Common.cs]: Got, it's right.

File: tests/Neo.Compiler.MSIL.UnitTests/TestClasses/Contract_TypeConvert.cs [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+            var bts1 = int1.AsByteArray();
+
+            var bts2=new byte[1] { 3 };
+            var int2 = bts2.AsBigInteger();
+
+            var bts3 = new byte[0];
+            var int3 = bts3.AsBigInteger();
+
+            var arrobj = new object[8];
+            arrobj[0] = int0;
```
Comment (Tommo-L) [inline tests/Neo.Compiler.MSIL.UnitTests/TestClasses/Contract_TypeConvert.cs]: I think we should add a test case for `_ConvertPushBoolean`.

Comment (lightszero) [reply tests/Neo.Compiler.MSIL.UnitTests/TestClasses/Contract_TypeConvert.cs]: That's hard,in IL sometimes bool is int too.

File: src/Neo.SmartContract.Framework/Helper.cs [hunk 6]
```diff
@@ -5,59 +5,62 @@ namespace Neo.SmartContract.Framework
...
+        //const string StackItemType_Array = "0x40";
+        //const string StackItemType_Struct = "0x41";
+        //const string StackItemType_Map = "0x48";
+        //const string StackItemType_InteropInterface = "0x60";
         /// <summary>
         /// Converts byte to byte[] considering the byte as a BigInteger (0x00 at the end)
         /// </summary>
-        [Script]
+        [OpCode(OpCode.CONVERT, StackItemType_ByteArray)]
         public extern static byte[] AsByteArray(this byte source);
```
Comment (erikzhang) [inline src/Neo.SmartContract.Framework/Helper.cs]: If we use `CONVERT`, we should rename the methods from `As***` to `To***`.

**Pull request merged**